File: src/main/kotlin/wowsft/model/gameparams/ship/component/atba/ATBA.java
package wowsft.model.gameparams.ship.component.atba;

import wowsft.config.WoWSFT;
import wowsft.model.gameparams.ship.component.airdefense.Aura;
import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Data;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;

@Data
@WoWSFT
@JsonIgnoreProperties(ignoreUnknown = true)
public class ATBA
{
    private List<Aura> auraFar = new ArrayList<>();
    private List<Aura> auraMedium = new ArrayList<>();
    private List<Aura> auraNear = new ArrayList<>();

    private List<Secondary> turrets = new ArrayList<>();
    private LinkedHashMap<String, Secondary> secondaries = new LinkedHashMap<>();

    private float maxDist;
    private float minDistH;
    private float minDistV;
    private float sigmaCount;
    private float taperDist;
    @JsonInclude
    private float GSIdealRadius = 1f;

    @JsonIgnore
    private ObjectMapper mapper = new ObjectMapper();

    @JsonAnySetter
    public void setGuns(String name, Object value)
    {
        if (value instanceof HashMap) {
            HashMap<String, Object> tempObject = mapper.convertValue(value, new TypeReference<HashMap<String, Object>>(){});

            if ("far".equalsIgnoreCase((String) tempObject.get("type"))) {
                auraFar.add(mapper.convertValue(value, Aura.class));
            } else if ("medium".equalsIgnoreCase((String) tempObject.get("type"))) {
                auraMedium.add(mapper.convertValue(value, Aura.class));
            } else if ("near".equalsIgnoreCase((String) tempObject.get("type"))) {
                auraNear.add(mapper.convertValue(value, Aura.class));
            } else if (tempObject.containsKey("HitLocationATBA")) {
                Secondary tempS = mapper.convertValue(value, Secondary.class);

                if (!secondaries.containsKey(tempS.getName())) {
                    tempS.setCount(1);
                    secondaries.put(tempS.getName(), tempS);
                } else {
                    secondaries.get(tempS.getName()).setCount(secondaries.get(tempS.getName()).getCount() + 1);
                }
            }
        }
    }
}

File: src/main/kotlin/wowsft/model/gameparams/ship/component/atba/ATBA.kt
package wowsft.model.gameparams.ship.component.atba

import wowsft.config.WoWSFT
import wowsft.model.gameparams.ship.component.airdefense.Aura
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper

import java.util.ArrayList
import java.util.HashMap
import java.util.LinkedHashMap

@WoWSFT
@JsonIgnoreProperties(ignoreUnknown = true)
class ATBA {
    var auraFar = ArrayList<Aura>()
    var auraMedium = ArrayList<Aura>()
    var auraNear = ArrayList<Aura>()

    var turrets = ArrayList<Secondary>()
    var secondaries = LinkedHashMap<String, Secondary>()

    var maxDist = 0.0
    var minDistH = 0.0
    var minDistV = 0.0
    var sigmaCount = 0.0
    var taperDist = 0.0
    @JsonInclude
    var GSIdealRadius = 1.0

    @JsonIgnore
    private val mapper = ObjectMapper()

    @JsonAnySetter
    fun setGuns(name: String, value: Any) {
        if (value is HashMap<*, *>) {
            val tempObject = mapper.convertValue<HashMap<String, Any>>(value, object : TypeReference<HashMap<String, Any>>() {})

            if ("far".equals(tempObject["type"] as String?, ignoreCase = true)) {
                auraFar.add(mapper.convertValue(value, Aura::class.java))
            } else if ("medium".equals(tempObject["type"] as String?, ignoreCase = true)) {
                auraMedium.add(mapper.convertValue(value, Aura::class.java))
            } else if ("near".equals(tempObject["type"] as String?, ignoreCase = true)) {
                auraNear.add(mapper.convertValue(value, Aura::class.java))
            } else if (tempObject.containsKey("HitLocationATBA")) {
                val tempS = mapper.convertValue(value, Secondary::class.java)

                if (secondaries[tempS.name] != null) {
                    secondaries[tempS.name]!!.count = secondaries[tempS.name]!!.count + 1
                } else {
                    tempS.count = 1
                    secondaries[tempS.name!!] = tempS
                }
            }
        }
    }
}
